[quant][pt2e] Add convert callback to Observer module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115001

Summary:
This is to allow easier extension of quant workflow in the future, as we are seening more
diverse ways of doing quantization

putting up this for feedbacks first

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_observer_callback

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add convert callback to Observer module

Summary:
This is to allow easier extension of quant workflow in the future, as we are seening more
diverse ways of doing quantization

putting up this for feedbacks first

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_observer_callback

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add convert callback to Observer module"


Summary:
This is to allow easier extension of quant workflow in the future, as we are seening more
diverse ways of doing quantization

putting up this for feedbacks first

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_observer_callback

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -105,10 +104,14 @@ def _replace_observer_with_quantize_dequantize_node_decomposed(
     assert isinstance(node.target, str)
     module_path, prefix = _get_module_path_and_prefix(node, node_name_to_scope, node_name_to_qconfig)
     activation_post_process = modules[node.target]
+    if hasattr(activation_post_process, "convert"):
+        activation_post_process.convert(model, node)
+        return
```
I would suggest filing a todo where we refactor all observers with convert method.

jerryzh168 replied:
yeah will do

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1712,3 +1712,106 @@ def test_reentrant(self):
+            def calculate_qparams(self, **kwargs):
+                pass
+
+            def convert(self, model: torch.fx.GraphModule, observer_node: Node):
```
I would prefer us to have `convert` do only specific things related to observer.

We need to figure out what this thing is?

My proposal is this:
Observer's forward method takes in a tensor and returns another tensor. Convert should just return equivalent of that but with whatever transforms they want to apply but not really doing any graph mutation.

e.g.
```
def convert(self, x):
   x_q = torch.ops.test_int4.quantiize_per_tensor_int4(...)
   return torch.ops.test_int4.dequantize_per_tensor_int4(x_q)
```

It is quant workflow's responsibility to replace original instance of observer with equivalent graph of convert. 

I would like to avoid 1) having observer to have think about both eager mode behavior (forward method) but also graph mode behavior for convert and 2) having to pass the entire GraphModule to convert which is error prone and allows for arbitrary mutation on the graph.

jerryzh168 replied:
yeah I thought about this, this is nice if we can achieve it, but I think it's hard to do because we also need to add new attributes to the graph module as well, what I wrote in the test was a simplest version that skipped a lot of details.

kimishpatel replied:
I think we should discuss in detail what we want to achieve here first though and then list constraints

kimishpatel replied:
I think we should do that first before landing. WDYT?

jerryzh168 replied:
yes sure, will follow up with you offline

kimishpatel reviewed (changes requested):
This is actually a pretty good feature to add, but I would like to make sure this is designed well

jerryzh168 commented:
@pytorchbot merge

jerryzh168 committed:
Update on "[quant][pt2e] Add convert callback to Observer module"


Summary:
This is to allow easier extension of quant workflow in the future, as we are seening more
diverse ways of doing quantization

putting up this for feedbacks first

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_observer_callback

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
